### ISSUE TITLE
TypeErorr in Grid Component when using area prop

### ISSUE THREAD

**Issue by vadistic** [grommet/grommet#2659]
Hello, I've noticed that Grid component is really prone to throwing TypeErrors. 

Ok, It works properly when used *exactly* as in Docs/Storybook, but there are many cases of TypeErrors happening when `areas` prop is used. It's nice-to-have level, but I thought it's worth reporting.

### Expected Behavior

No TypeErrors when providing props of valid type.

Maybe some input sanitation of props (typeguards?) and info in docs.

### Actual Behavior
Let's take Grid Component:
```tsx
<Grid
  gap="medium"
  // areas must be defined for bug to happen :)
  areas={[{ name: "image", start: [0, 0], end: [0, 1] }]}
  columns={["small", "flex", "small"]}
  rows={["flex", "medium"]} // Works fine
>
  {/* ... */}
</Grid>

<Grid
  gap="medium"
  areas={[{ name: "image", start: [0, 0], end: [0, 1] }]}
  columns={["small", "flex", "small"]}
  rows={undefined} // TypeError "Cannot read property 'map' of undefined"
>
  {/* ... */}
</Grid>

<Grid
  gap="medium"
  areas={[{ name: "image", start: [0, 0], end: [0, 1] }]}
  columns={["small", "flex", "small"]}
  rows={"flex"} // TypeError "props.rowsProp.map is not a function"
>
  {/* ... */}
</Grid>

<Grid
  gap="medium"
  areas={[{ name: "image", start: [0, 0], end: [0, 1] }]}
  columns={["small", "flex", "small"]}
  rows={["flex"]} // TypeError: "Cannot set property '0' of undefined"
>
  {/* ... */}
</Grid>
```

### URL, screen shot, or Codepen exhibiting the issue

CODESANDBOX: https://codesandbox.io/s/m703y9729j

### Steps to Reproduce

Uncomment stuff in codesandbox

### Bonus
- Grid component typescript typings for `columns` props are missing case like `(keyof Size)[][]` allowing only `"auto"[][]`
- Why the heck background color does not work on that box there in codesandbox?^^

### Your Environment

- Grommet version: 2.3.1
- Browser Name and version: Chrome 72 (Official Beta)
- Operating System and version: Linux/ whatever they run on CodeSandbox

**Comment by ShimiSun** [comment]
Many interesting comments @vadistic.
I took a look at the codesandbox example, and it was not clear what you are trying to design since there was no compatibility with the (# of rows) to the (# of columns) as it was described it `areas` definition.

Let's take for example:
```<Grid
  gap="medium"
  areas={[{ name: "image", start: [0, 0], end: [0, 1] }]}
  columns={["small", "flex", "small"]}
  rows={["flex"]} // TypeError "props.rowsProp.map is not a function"
>
  {/* ... */}
</Grid>
```
The `areas` are defining only one column and two rows, so the input of just one row is not compatible, and the calculation failed. It does work if you define only one column and two rows though. since it will be compatible with the `areas` defined. see https://codesandbox.io/s/j3n9oxxr43
Another option will be to change the areas definition and then `rows={["flex"]}` will work:
https://codesandbox.io/s/7ww6z647y6

I do want to imprve the documentation, so if you feel like some information was missing for you to understand how this component should work, please share your thoughts.

Regarding the example of `rows={"flex"}` we are providing a warning of
```
`Grid `areas` requires `rows` and 'columns' to be arrays.`  
```
so this will not work unless you use an array for `rows`. I agree we might need to document it better.

Regarding the background color issue, you are correct, you provided the `theme` as expected, but there is an issue with the `react-dom` version in the sandbox. If you'll upgrad its version to 16.7.0 it will work. we are working on this issue.

Typescript - yea its a bug, we'll track it with this issue number or feel free to file a new issue.

hit me back and let me know what you think.

**Comment by MikeWalrath** [comment]
@vadistic the above is wonderful collaboration. Do you have any additional thoughts on the issue?

**Comment by vadistic** [comment]
Thanks for reminder @MikeWalrath, sorry for leaving it hanging for so long, I've changed my focus for a bit.

@ShimiSun you've addresed my concers, the main problem I had appears to be my overlook of  `[0,1] !== [1,0]` and my areas did not match :man_facepalming: 

The warning of array requirement seems to be reasonable solution for me, I did not notice it under crash call stack the first time and got a bit misled by :
> Specifying a single string will repeat multiple columns of that size, as long as there is room for more. 

One liner in docs could be tiny improvement, but frankly I'm not sure if necessary with this console warning. Maybe simply allowing string value instead would be good UX.

And for typescript - I see it's still as it were so I'm making quick PR and we can close this issue :)
https://github.com/grommet/grommet/pull/2782
